api/v1beta1/common_types.go

```diff
@@ -126,6 +126,10 @@ const (
 	// any changes to the actual object because it is a dry run) and the topology controller
 	// will receive the resulting object.
 	TopologyDryRunAnnotation = "topology.cluster.x-k8s.io/dry-run"
+
+	// ReplicasManagedByAutoscalerAnnotation is an annotation that indicates an external autoscaler controls replica count.
+	// This annotation can be used to inform MachinePool status during in-progress scaling scenarios.
+	ReplicasManagedByAutoscalerAnnotation = "cluster.x-k8s.io/replicas-managed-by-autoscaler"
```

Comment:
By reading "ReplicasManagedByAutoscalerAnnotation" I would think this has something to do with [capi cluster autoscaler provider](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler/cloudprovider/clusterapi), which is not true.

Would it be something like "ReplicasExternallyManaged" more appropriate? and since this concept only apply to MachinePools "MachinePoolReplicasExternallyManagedAnnotation"?

Thoughts?

Reply:
That's fair, I could see people getting confused and thinking this has to do with the CAPI autoscaler being enabled. 

How about: `cluster.x-k8s.io/replicas-managed-by-external-autoscaler`?

`cluster.x-k8s.io/replicas-externally-managed` works too, I just think mentioning an autoscaler in there is a bit more specific as otherwise it could imply the machinepool machines themselves are not managed by CAPI which isn't the case.

IMO adding MachinePools in there makes it become a bit too much of a mouthful `cluster.x-k8s.io/machine-pool-replicas-managed-by-external-autoscaler`, and is probably redundant anyways since the annotation will be applied on the MachinePool object name itself (but maybe worth putting it in the constant name so it's clear where this should be used).

So TLDR, how about:

```
// MachinePoolReplicasManagedByExternalAutoscalerAnnotation is an annotation that indicates an external autoscaler controls replica count.
// This annotation can be used to inform MachinePool status during in-progress scaling scenarios.
MachinePoolReplicasManagedByExternalAutoscalerAnnotation = "cluster.x-k8s.io/replicas-managed-by-external-autoscaler"
```

Reply:
I agree that the word "autoscaler" helps to disambiguate from potential "nodes are externally managed" functionality. The key difference is disambiguating "nodes" from "replicas", and I think having the "autoscaler" mention in there helps solidify that we are only talking about the replica count being out of cluster-api's enforcement scope.

Reply:
I also think that this annotation could (and IMO should) be re-used by MachineDeployments, and is not MachinePool-specific.

Reply:
I'm good with any of your suggestions above. Just sharing some more ideas for consideration in case helps you to make the final choice:

Have you consider reusing https://github.com/kubernetes-sigs/cluster-api/blob/main/api/v1beta1/common_types.go#L113-L119?

>I agree that the word "autoscaler" helps to disambiguate from potential "nodes are externally managed" functionality. The key difference is disambiguating "nodes" from "replicas", and I think having the "autoscaler" mention in there helps solidify that we are only talking about the replica count being out of cluster-api's enforcement scope.

When I read replicas-managed-by-external-autoscaler my first thought is some external autoscaler manages .replicas to scale in/out. But what is happening is something creates cloud instances and reconciles .replicas to make sure it's consistent with reality as if it were .status.

>I also think that this annotation could (and IMO should) be re-used by MachineDeployments, and is not MachinePool-specific.

Agree, but when/if there's a use case for it.

Reply:
I'm not sure we could re-use the `cluster.x-k8s.io/managed-by` annotation as that suggests the whole resource is managed by something else (and CAPx should therefore ignore it). There's no way to scope that annotation to the `.replicas` property.

Reply:
I agree there is some ambiguity here:

- Is the autoscaler doing the horizontal scaling at the infra level (below the abstraction layer of capi)?
- Is the autoscaler operating directly against capi, but is the only allowed client to do so?

I am advocating for an explicit annotation that communicates the first condition only. I'm not sure an annotation is an appropriate actor in the 2nd scenario, and the changes to the way we calculate an appropriate status message for the detection of drift between capi's replica count and the observed number of ready nodes/machines is also only to address the first scenario, when capi is no longer authoritative for replica count, but is instead passively consuming it from infra state.

Because we've been successfully prototyping `"cluster.x-k8s.io/replicas-managed-by-autoscaler"` in capz for a while, I'd advocate we stick with that (and in lieu of any consensus in the progression towards an alternative over the past few months).

Reply:
this goes back to the comment @enxebre made that started this discussion

> By reading "ReplicasManagedByAutoscalerAnnotation" I would think this has something to do with [capi cluster autoscaler provider](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler/cloudprovider/clusterapi), which is not true.

it's really not intuitive that this doesn't refer to the CAPI autoscaler being enabled. Why not `cluster.x-k8s.io/replicas-managed-by-external-autoscaler`?

Reply:
This is a fair observation, I've made that change, can folks following this PTAL and comment on if we want to continue reconsidering a different annotation key string? Thanks!